fetch-in-react-practice/client-ts/src/components/Todos.tsx

```diff
@@ -0,0 +1,162 @@
+/* eslint-disable @typescript-eslint/no-unused-vars -- Remove me */
+import React, { useEffect, useState } from 'react';
+import PageTitle from './PageTitle';
+import TodoList from './TodoList';
+import TodoForm from './TodoForm';
+
+export type UnsavedTodo = {
+  task: string;
+  isCompleted: boolean;
+};
+export type Todo = UnsavedTodo & {
+  todoId: number;
+};
+
+export default function Todos() {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<unknown>();
+
+  /* Implement useEffect to fetch all todos. Hints are at the bottom of the file. */
+  useEffect(() => {
+    async function fetchTodos() {
+      try {
+        const res = await fetch('/api/todos');
+        if (!res.ok) {
+          throw new Error(`fetch Error ${res.status}`);
+        }
+        const todosList: Todo[] = await res.json();
+        setTodos(todosList);
+      } catch (err) {
+        setError(err);
+      } finally {
+        setIsLoading(false);
+      }
+    }
+    fetchTodos();
+  }, []);
+
+  /* Implement addTodo to add a new todo. Hints are at the bottom of the file. */
+  async function addTodo(newTodo: UnsavedTodo) {
+    try {
+      const options = {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(newTodo),
+      };
+      const res = await fetch('/api/todos', options);
+      if (!res.ok) {
+        throw new Error(`fetch Error ${res.status}`);
+      }
+      const savedTodo: Todo = await res.json();
+      setTodos((prev) => prev.concat(savedTodo));
+    } catch (err) {
+      setError(err);
+    } finally {
+      setIsLoading(false);
```

Comment:
The way this UI works, this isn't really needed since it will already be false from the useEffect. If you were using a spinner or some other way to indicate the operation was ongoing, you would want to `setIsLoading(true)` at the beginning of this function, then set it false in the `finally` like you've done.

If you're bored, this would be a good exercise for you -- figure out how to indicate the operation is in progress and set isLoading appropriately. Perhaps the best UI I've seen for this is:
* When user clicks the Add button you disable and dim the button (so further clicks don't accidentally resend the request)
* While the operation is ongoing, draw a spinner inside the button (alternatively, spin the cursor or draw a spinner in the middle of the screen with the rest of the screen dimmed)

To test this, you can use Developer Tools to slow down the network, or you can go into the server and introduce a pause in the `add` endpoint. A nice way to do that is:
```
async function delay(msecs: number): Promise<void> {
    return new Promise((resolve) => setTimeout(() => resolve(), msecs));
}

await delay(1000);
```